File: entities/weeping_block.py
# entities/weeping_block.py
import math
from .base_entity import Entity
from settings import *


class WeepingBlock(Entity):
    def __init__(self, game, x, y):
        super().__init__(game, x, y, (120, 120, 120), "circle")
        self.speed = 6.0
        self.active = False

    def update(self):
        super().update()

        px, py = self.game.player.pos
        p_angle = self.game.player.angle

        dx = self.x - px
        dy = self.y - py
        dist_to_player = math.sqrt(dx * dx + dy * dy)

        angle_to_entity = math.atan2(dy, dx) - p_angle
        while angle_to_entity > math.pi: angle_to_entity -= 2 * math.pi
        while angle_to_entity < -math.pi: angle_to_entity += 2 * math.pi

        player_is_looking = -HALF_FOV < angle_to_entity < HALF_FOV

        if player_is_looking:
            self.color = (120, 120, 120)
            self.active = False
        else:
            self.color = (200, 200, 200)
            self.active = True

            if dist_to_player > 30:
                target_angle = math.atan2(py - self.y, px - self.x)
                self.angle = target_angle
                self.move(math.cos(self.angle) * self.speed, math.sin(self.angle) * self.speed)

File: src/entities/weeping_block.rs
// src/entities/weeping_block.rs
use std::f32::consts::PI;
use crate::config::*;
use crate::map::MapHandler;
use super::entity_base::EntityBase;

pub fn update_weeping_block(
    base: &mut EntityBase,
    active: &mut bool,
    player_pos: (f32, f32),
    player_angle: f32,
    map: &mut MapHandler,
    dt: f32,
) {
    let speed = base.speed * dt * 60.0;
    let (px, py) = player_pos;
    let dx = base.x - px;
    let dy = base.y - py;
    let dist = (dx * dx + dy * dy).sqrt();
    let mut angle_to = dy.atan2(dx) - player_angle;
    while angle_to > PI { angle_to -= 2.0 * PI; }
    while angle_to < -PI { angle_to += 2.0 * PI; }
    let looking = angle_to.abs() < HALF_FOV;

    if looking {
        base.color = [100.0/255.0, 100.0/255.0, 100.0/255.0];
        *active = false;
    } else {
        base.color = [220.0/255.0, 220.0/255.0, 200.0/255.0];
        *active = true;
        if dist > 30.0 {
            let target = (py - base.y).atan2(px - base.x);
            base.angle = target;
            base.move_pos(
                base.angle.cos() * speed,
                base.angle.sin() * speed,
                map,
            );
        }
    }
}
